src/processTargets/modifiers/scopeHandlers/scopeHandler.types.ts

```diff
@@ -0,0 +1,194 @@
+//    * support things like "every token block", "token" will have a secondary
+//    * iteration scope of the entire file.
+//    *
+//    * It is common to either return a single iteration scope, or to return a
+//    * list of two iteration scopes: the preferred scope, followed by an
+//    * iteration scope representing the entire file.  Returning more than these
+//    * two is just an optimization.
+//    * @param position The position from which to expand
+//    */
+//   getTargetsIntersectingRange(editor: TextEditor, range: Range): Target[];
+
+//   getTargetsInIterationScopeContainingRange(
+//     editor: TextEditor,
+//     range: Range
+//   ): Target[];
+// },
+
+export interface ContainedIndices {
+  start: number;
+  end: number;
+}
+
+export interface IterationScope {
+  targets: Target[];
+  containingIndices: ContainedIndices | undefined;
+}
+
+export abstract class ScopeHandler {
+  constructor(protected scopeType: ScopeType) {}
+
+  run(
+    editor: TextEditor,
+    contentRange: Range,
+    isReversed: boolean,
+    hasExplicitRange: boolean
+  ): IterationScope {
+    const targets = this.getEveryTarget(
+      editor,
+      contentRange,
+      isReversed,
+      hasExplicitRange
+    );
+
+    const containingIndices = contentRange.isEmpty
+      ? this.getContainingIndicesForPosition(contentRange.start, targets)
+      : this.getContainingIndicesForRange(contentRange, targets);
+
+    return {
+      targets,
+      containingIndices,
+    };
+  }
+
+  protected getContainingIndicesForPosition(
+    position: Position,
+    targets: Target[]
+  ): ContainedIndices | undefined {
+    const mappings = targets
+      .map((target, index) => ({ range: target.contentRange, index }))
+      .filter((mapping) => mapping.range.contains(position));
+
+    if (mappings.length === 0) {
+      return undefined;
+    }
+
+    const index = mappings.at(-1)!.index;
+
+    return { start: index, end: index };
+  }
+
+  protected getContainingIndicesForRange(
+    range: Range,
+    targets: Target[]
+  ): ContainedIndices | undefined {
+    const mappings = targets
+      .map((target, index) => ({ range: target.contentRange, index }))
+      .filter((mapping) => {
+        const intersection = mapping.range.intersection(range);
+        return intersection != null && !intersection.isEmpty;
+      });
+
+    if (mappings.length === 0) {
+      return undefined;
+    }
+
+    return { start: mappings[0].index, end: mappings.at(-1)!.index };
+  }
+
+  protected filterRangesByIterationScope(
+    iterationScope: Range,
+    ranges: Range[]
+  ): Range[] {
+    return ranges.filter((r) => {
+      const intersection = r.intersection(iterationScope);
+      return intersection != null && !intersection.isEmpty;
+    });
+  }
+
+  protected abstract getEveryTarget(
+    editor: TextEditor,
+    contentRange: Range,
+    isReversed: boolean,
+    hasExplicitRange: boolean
```

Comment:
I'm not sure this is what we want to be passing in here; won't "next" want to fake it to `false`?  I think it's confusing to have something that is an attribute of the target and then to hack it to be false.  Could we maybe just name the variable something else?

I think it would be useful to see "next" implemented, as this impl is still doing the hacky faked explicit range thing via calling "every".  Shouldn't be too hard, no?

Reply:
Yeah probably. As I said at the top names could probably be improved. 

I can do that. I'm just a bit has content of having these different implementations: new/legacy without support from the type system.

Reply:
Ok now relative and ordinal scope stages use the new scope handler for token